### DISCUSSION TITLE
Add salt to password's hash

### DISCUSSION THREAD

**Discussion by aviramha** [fastapi-users/fastapi-users#776] [Ideas]
Right now the password hash isn't salted, which means that if an evil actor gains access to the database they can use rainbow table to reverse the plaintext.

**Comment by frankie567** [comment]
The password hash *is* salted. Everything is handled by Passlib using the BCrypt algorithm: 

https://github.com/fastapi-users/fastapi-users/blob/e0e8dfbc3b2af100863bbe542168cbc48bceae85/fastapi_users/password.py#L6-L16

Ref: https://passlib.readthedocs.io/en/stable/lib/passlib.hash.bcrypt.html

**Reply by aviramha** [reply]
How would it work then if the salt is automatically generated? The salt would reset between instances..

**Reply by aviramha** [reply]
Ah, the salt is per password. Sorry, my bad!

**Reply by frankie567** [reply]
The salt is concatenated with the hash and stored in the same field. You can check technical details on [BCrypt on Wikipedia
](https://en.bitcoinwiki.org/wiki/Bcrypt):

> The bcrypt function is the default password hash algorithm for OpenBSD and other systems including some Linux distributions such as SUSE Linux. The prefix "$2a$" or "$2b$" (or "$2y$") in a hash string in a shadow password file indicates that hash string is a bcrypt hash in modular crypt format. The rest of the hash string includes the cost parameter, a 128-bit salt (Radix-64 encoded as 22 characters), and 184 bits of the resulting hash value (Radix-64 encoded as 31 characters). The Radix-64 encoding uses the unix/crypt alphabet, and is not 'standard' Base-64. The cost parameter specifies a key expansion iteration count as a power of two, which is an input to the crypt algorithm.
> 
> For example, the shadow password record $2a$10$N9qo8uLOickgx2ZMRZoMyeIjZAgcfl7p92ldGxad68LJZdL17lhWy specifies a cost parameter of 10, indicating 210 key expansion rounds. The salt is N9qo8uLOickgx2ZMRZoMye and the resulting hash is IjZAgcfl7p92ldGxad68LJZdL17lhWy. Per standard practice, the user's password itself is not stored.

**Reply by frankie567** [reply]
> Ah, the salt is per password. Sorry, my bad!

Yes! That's why it's super safe against rainbow tables 😉
